File: src/main/java/me/imoltres/bbu/game/GameProgression.java
package me.imoltres.bbu.game;

import lombok.Getter;
import lombok.Setter;
import me.imoltres.bbu.data.player.BBUPlayerStatistics;
import me.imoltres.bbu.data.team.BBUTeam;
import org.jetbrains.annotations.NotNull;

import java.util.HashSet;
import java.util.Set;
import java.util.stream.Collectors;

public class GameProgression {

    public final Set<BBUPlayerStatistics> playerStatistics = new HashSet<>();
    public final Set<BBUTeam> teamsAlive = new HashSet<>();
    @Getter
    @Setter
    @NotNull
    private GameState gameState = GameState.PRE_GAME;

    public Set<BBUTeam> getTeams(boolean hasBeacon) {
        return teamsAlive
                .stream()
                .filter(team -> hasBeacon == team.hasBeacon())
                .collect(Collectors.toSet());
    }

}

File: src/main/java/me/imoltres/bbu/game/GameProgression.kt
package me.imoltres.bbu.game

import me.imoltres.bbu.BBU
import me.imoltres.bbu.data.team.BBUTeam
import me.imoltres.bbu.utils.config.MainConfig
import java.util.stream.Collectors

class GameProgression {
    var gameState = GameState.LOBBY
    var currentShrinkPhase: ShrinkPhase? = null

    private var shrinkIndex = -1

    fun getOrderedShrinkPhases(): List<ShrinkPhase> {
        return MainConfig.borderPhases.sortedBy { it.time }
    }

    fun getTeams(hasBeacon: Boolean): Set<BBUTeam> {
        return getTeamsAlive()
            .stream()
            .filter { team: BBUTeam -> hasBeacon == team.hasBeacon() }
            .collect(Collectors.toSet())
    }

    fun getTeamsAlive(): Set<BBUTeam> {
        return BBU.getInstance().teamController.allTeams.filter { bbuTeam -> bbuTeam.hasBeacon() || bbuTeam.players.isNotEmpty() }
            .toSet()
    }

    fun getNextPhase(): ShrinkPhase? {
        val orderedShrinkPhases = getOrderedShrinkPhases()
        if (shrinkIndex + 1 >= orderedShrinkPhases.size) return null

        return orderedShrinkPhases[shrinkIndex + 1]
    }

    fun getPreviousPhase(): ShrinkPhase? {
        val orderedShrinkPhases = getOrderedShrinkPhases()
        if (shrinkIndex - 1 >= orderedShrinkPhases.size) return null // last index, no next phase
        if (shrinkIndex - 1 < 0) return null                         // first index, no previous phase

        return orderedShrinkPhases[shrinkIndex - 1]
    }

}